### DISCUSSION TITLE
[New Feature] Run strands safely on a separate thread

### DISCUSSION THREAD

**Discussion by lasinicl** [ballerina-platform/ballerina-lang#36307] [General]
$title will be achieved by

1. defining a `start-action` to be isolated
2. extending isolated concept to named workers

 **Defining a `start-action` to be isolated** 

A `start-action` will be defined isolated, if
- the method or function being called is isolated
- the expression for every argument is an isolated expression
- for a `method-call` or `client-remote-method-call-action` the expression for the object is an isolated expression

A `start-action` is modeled as an action invocation in the compiler. `ISOLATED` flag will be added for the invocation at the IsolationAnayzer if it satisfies the above conditions, so that can be identified as an isolated `start-action`.

The strand created by an isolated `start-action` can run in a different thread from the current strand - `@strand{thread:"any"}` annotation will be added internally for the isolated `start-action` at the desugar phase. 

An isolated function can contain an isolated `start-action` - According to the current implementation, it is not allowed to have `start-action` in the body of an isolated function. But with this change, it will be allowed to have an isolated `start-action` in an isolated function and the compilation errors will be updated to reflect the invalid invocations.

**Extending isolated concept to named workers**

An isolated function can contain a `named-worker-decl` - According to the current implementation, it is not allowed to have a `named-worker-decl` in the body of an isolated function. But with this change, it will be allowed to have an isolated worker in an isolated function and the compilation errors will be updated to reflect the invalid worker declarations.

When a named worker is isolated its body will have similar restrictions to an isolated function (https://ballerina.io/spec/lang/master/#isolated_functions). Within a `named-worker-decl` in an isolated function, a reference to a local variable outside the `named-worker-decl` should satisfy the same requirements as a captured variable in an isolated function/method.

> Captured variables are restricted as follows: if the body is part of an anonymous-function-expr, then any captured variable references must refer to variables that are both final (either explicitly or implicitly) and have a static type that is a subtype of readonly|isolated object {}.

A worker is modeled as a lambda function in the compiler. `ISOLATED` flag will be added for the lambda function at the IsolationAnayzer if it satisfies the above conditions, so that can be identified as an isolated worker.

The named worker's strand will run on a separate thread from the function's strand - `@strand{thread:"any"}` annotation will be added internally for the isolated worker at the desugar phase. 

**Isolated inference for `start-action` and named workers**
Should also support inferring the isolatedness of a `start-action` and a named worker. 

Highly appreciate your input on this discussion.

**Comment by warunalakshitha** [comment]
@lasinicl If the code contains @strand{thread:"any"} to a non isolated start action or named worker will that be a compilation error?
And also if code has @strand{thread:"parent"} in an isolated start action or named worker should we still run them on separate thread?

**Reply by lasinicl** [reply]
> If the code contains @strand{thread:"any"} to a non isolated start action or named worker will that be a compilation error? 

An isolated function can not contain a non-isolated start action or a named worker. Hence it will be a compilation error. 
A non-isolated function can contain a non-isolated start action or a named worker. We can use the annotations as well.

> If code has @strand{thread:"parent"} in an isolated start action or named worker should we still run them on a separate thread

If so, we will be ignoring the `@strand{thread: "parent"}` annotation and add the `@strand{thread: "any"}` annotation internally so that it runs on a separate thread. This was decided since it's currently in discussion to deprecate the `@strand{}` annotation. We will also be logging a warning if `@strand{thread: "parent"}` has been used in such cases.

**Reply by lasinicl** [reply]
Update: It will be an error to use @strand on a start-action or named worker that occurs in a function explicitly declared as isolated. (https://github.com/ballerina-platform/ballerina-spec/issues/1068#issuecomment-1139472106)

**Comment by lasinicl** [comment]
When `@strand` is used on a start action or a named worker, 

- if the containing function is explicitly declared as isolated, it will be a compile-time error 
- otherwise
        - the containing function will not be inferred to be isolated 
        - the existing `@strand` runtime semantics will continue as now
        - there will be a deprecation warning saying not to use `@strand`, but instead use the safe features introduced by this issue
